File: src/main/java/com/gtladd/gtladditions/common/machine/muiltblock/controller/TaixuTurbidArray.java
package com.gtladd.gtladditions.common.machine.muiltblock.controller;

import org.gtlcore.gtlcore.api.machine.multiblock.ParallelMachine;
import org.gtlcore.gtlcore.api.pattern.util.IValueContainer;
import org.gtlcore.gtlcore.common.data.GTLMaterials;
import org.gtlcore.gtlcore.common.machine.multiblock.electric.TierCasingMachine;
import org.gtlcore.gtlcore.utils.Registries;

import com.gregtechceu.gtceu.api.GTValues;
import com.gregtechceu.gtceu.api.block.ICoilType;
import com.gregtechceu.gtceu.api.capability.recipe.FluidRecipeCapability;
import com.gregtechceu.gtceu.api.capability.recipe.IO;
import com.gregtechceu.gtceu.api.gui.GuiTextures;
import com.gregtechceu.gtceu.api.machine.IMachineBlockEntity;
import com.gregtechceu.gtceu.api.machine.MetaMachine;
import com.gregtechceu.gtceu.api.machine.feature.IMachineModifyDrops;
import com.gregtechceu.gtceu.api.machine.multiblock.WorkableMultiblockMachine;
import com.gregtechceu.gtceu.api.machine.trait.NotifiableItemStackHandler;
import com.gregtechceu.gtceu.api.pattern.util.PatternMatchContext;
import com.gregtechceu.gtceu.api.recipe.GTRecipe;
import com.gregtechceu.gtceu.api.recipe.RecipeHelper;
import com.gregtechceu.gtceu.api.recipe.logic.OCParams;
import com.gregtechceu.gtceu.api.recipe.logic.OCResult;
import com.gregtechceu.gtceu.common.block.CoilBlock;
import com.gregtechceu.gtceu.common.data.GTMaterials;
import com.gregtechceu.gtceu.common.data.GTRecipeModifiers;
import com.gregtechceu.gtceu.common.data.GTRecipeTypes;
import com.gregtechceu.gtceu.data.recipe.builder.GTRecipeBuilder;
import com.gregtechceu.gtceu.utils.GTUtil;

import com.lowdragmc.lowdraglib.gui.widget.SlotWidget;
import com.lowdragmc.lowdraglib.gui.widget.Widget;
import com.lowdragmc.lowdraglib.gui.widget.WidgetGroup;
import com.lowdragmc.lowdraglib.misc.ItemStackTransfer;
import com.lowdragmc.lowdraglib.syncdata.annotation.Persisted;
import com.lowdragmc.lowdraglib.syncdata.field.ManagedFieldHolder;
import com.lowdragmc.lowdraglib.utils.Size;

import net.minecraft.MethodsReturnNonnullByDefault;
import net.minecraft.network.chat.Component;
import net.minecraft.resources.ResourceLocation;
import net.minecraft.world.item.Item;
import net.minecraft.world.item.ItemStack;

import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.util.*;

import javax.annotation.ParametersAreNonnullByDefault;

@ParametersAreNonnullByDefault
@MethodsReturnNonnullByDefault
public class TaixuTurbidArray extends TierCasingMachine implements ParallelMachine, IMachineModifyDrops {

    public static final ManagedFieldHolder MANAGED_FIELD_HOLDER;
    @Persisted
    public final NotifiableItemStackHandler machineStorage;
    private ICoilType coilType;
    private int height = 0;
    private static final Map<Integer, Integer> coil = new HashMap<>();
    private static final ItemStack ENDERIUM = Registries.getItemStack("gtceu:enderium_nanoswarm", 64);
    private static final ItemStack DRACONIUM = Registries.getItemStack("gtceu:draconium_nanoswarm", 64);
    private static final ItemStack SPACETIME = Registries.getItemStack("gtceu:spacetime_nanoswarm", 64);
    private static final ItemStack ETERNITY = Registries.getItemStack("gtceu:eternity_nanoswarm", 64);

    public TaixuTurbidArray(IMachineBlockEntity holder) {
        super(holder, "SCTier");
        this.machineStorage = createMachineStorage();
        this.coilType = CoilBlock.CoilType.CUPRONICKEL;
    }

    protected NotifiableItemStackHandler createMachineStorage() {
        NotifiableItemStackHandler handler = new NotifiableItemStackHandler(this, 1, IO.NONE, IO.BOTH, (slots) -> new ItemStackTransfer(1) {

            public int getSlotLimit(int slot) {
                return 64;
            }
        });
        handler.setFilter(this::filter);
        return handler;
    }

    @Override
    public @NotNull Widget createUIWidget() {
        Widget widget = super.createUIWidget();
        if (widget instanceof WidgetGroup group) {
            Size size = group.getSize();
            group.addWidget((new SlotWidget(this.machineStorage.storage, 0, size.width - 30, size.height - 30, true, true))
                    .setBackground(GuiTextures.SLOT).setHoverTooltips(this.slotTooltips()));
        }
        return widget;
    }

    protected boolean filter(@NotNull ItemStack itemStack) {
        Item item = itemStack.getItem();
        return ENDERIUM.is(item) || DRACONIUM.is(item) || SPACETIME.is(item) || ETERNITY.is(item);
    }

    public void onDrops(List<ItemStack> drops) {
        this.clearInventory(this.machineStorage.storage);
    }

    private List<Component> slotTooltips() {
        List<Component> tooltip = new ArrayList<>();
        tooltip.add(Component.literal("最多可以放入64个物品"));
        tooltip.add(Component.literal("可放入以下物品与提供对应的加成："));
        tooltip.add(Component.literal("末影纳米蜂群：0.01"));
        tooltip.add(Component.literal("龙纳米蜂群：0.05"));
        tooltip.add(Component.literal("时空纳米蜂群：0.1"));
        tooltip.add(Component.literal("永恒纳米蜂群：0.2"));
        return tooltip;
    }

    public static @Nullable GTRecipe recipeModifier(MetaMachine machine, @NotNull GTRecipe recipe, @NotNull OCParams params, @NotNull OCResult result) {
        if (machine instanceof TaixuTurbidArray turbidArray) {
            GTRecipe recipe1 = recipe.copy();
            GTRecipeBuilder builder = new GTRecipeBuilder(Objects.requireNonNull(ResourceLocation.tryParse("uu")), GTRecipeTypes.DUMMY_RECIPES);
            recipe1.outputs.put(FluidRecipeCapability.CAP, new ArrayList<>());
            if (Math.random() * 100 <= turbidArray.successRateA() && turbidArray.getEnergyTier() >= GTValues.UIV) {
                builder.outputFluids(GTLMaterials.UuAmplifier.getFluid(turbidArray.baseOutputFluid1()));
            }
            if (Math.random() * 100 <= turbidArray.successRateB() && turbidArray.getEnergyTier() >= GTValues.OpV) {
                builder.outputFluids(GTMaterials.UUMatter.getFluid(turbidArray.baseOutputFluid2()));
            }
            if (builder.buildRawRecipe().outputs.containsKey(FluidRecipeCapability.CAP)) {
                recipe1.outputs.get(FluidRecipeCapability.CAP).addAll(builder.buildRawRecipe().outputs.get(FluidRecipeCapability.CAP));
            }
            recipe1 = GTRecipeModifiers.accurateParallel(machine, recipe1, turbidArray.getMaxParallel(), false).getFirst();
            recipe1.duration = 100;
            RecipeHelper.setInputEUt(recipe1, 524288L * GTValues.V[turbidArray.getEnergyTier()]);
            return recipe1;
        }
        return null;
    }

    public void onStructureFormed() {
        super.onStructureFormed();
        PatternMatchContext context = this.getMultiblockState().getMatchContext();
        Object type = context.get("CoilType");
        if (type instanceof ICoilType coiltype) this.coilType = coiltype;
        Object var3 = context.getOrCreate("SpeedPipeValue", IValueContainer::noop).getValue();
        if (var3 instanceof Integer integer) this.height = integer - 2;
    }

    public boolean beforeWorking(@Nullable GTRecipe recipe) {
        return true;
    }

    public void addDisplayText(@NotNull List<Component> textList) {
        super.addDisplayText(textList);
        if (this.isFormed) {
            textList.add(Component.literal("高度：" + this.height));
            textList.add(Component.literal("最大并行数：" + this.getMaxParallel()));
            if (this.getEnergyTier() >= GTValues.UIV) {
                textList.add(Component.literal("UU增幅液成功概率：" + this.successRateA() + "%"));
                textList.add(Component.literal("UU增幅液基础输出量：" + this.baseOutputFluid1() + "mb"));
                if (this.getEnergyTier() >= GTValues.OpV) {
                    textList.add(Component.literal("UU物质成功概率：" + this.successRateB() + "%"));
                    textList.add(Component.literal("UU物质基础输出量：" + this.baseOutputFluid2() + "mb"));
                }
            }
        }
    }

    private double frameA() {
        return 8.0 * (Math.pow(2, this.getCasingTier()) - 1) * Math.sqrt(GTValues.ALL_TIERS[this.getEnergyTier()] + 1);
    }

    private double frameB() {
        return 3.8 * Math.pow(1.3, coil.get(this.coilType.getCoilTemperature())) * Math.pow(this.coilType.getCoilTemperature() / 36000.0, 0.7);
    }

    private int successRateA() {
        return (int) Math.round(100 / (1 + Math.exp(-0.1 * (this.frameA() / 50 + this.frameB() / 100 + this.height / 9.0))) + this.getSlotAdd());
    }

    private int successRateB() {
        return (int) Math.round(100 * (1 - Math.exp(-0.02 * ((this.frameA() + this.frameB()) / 20.0 + Math.cbrt(this.height) * this.getEnergyTier() / 7.0))) + this.getSlotAdd());
    }

    private int baseOutputFluid1() {
        return (int) (4096 * (1 - Math.exp(-0.015 * (this.frameA() * this.height / 16.0 + this.frameB() * Math.log(this.getEnergyTier() + 2)))));
    }

    private int baseOutputFluid2() {
        return (int) (2250 * Math.tanh(Math.sqrt(this.frameA() * this.frameB()) * (this.height + this.getEnergyTier()) * 0.06 / 200.0));
    }

    @Override
    public int getMaxParallel() {
        return (int) (4096 * Math.pow(1.5, ((double) this.coilType.getCoilTemperature() / 6400)));
    }

    private int getEnergyTier() {
        return GTUtil.getFloorTierByVoltage(this.getMaxVoltage());
    }

    private double getSlotAdd() {
        Item item = this.machineStorage.storage.getStackInSlot(0).getItem();
        int amount = this.machineStorage.storage.getStackInSlot(0).getCount();
        if (ENDERIUM.is(item)) return 0.01 * amount;
        else if (DRACONIUM.is(item)) return 0.05 * amount;
        else if (SPACETIME.is(item)) return 0.1 * amount;
        else if (ETERNITY.is(item)) return 0.2 * amount;
        return 0.0;
    }

    public @NotNull ManagedFieldHolder getFieldHolder() {
        return MANAGED_FIELD_HOLDER;
    }

    static {
        MANAGED_FIELD_HOLDER = new ManagedFieldHolder(TaixuTurbidArray.class, WorkableMultiblockMachine.MANAGED_FIELD_HOLDER);
        coil.put(1800, 1);
        coil.put(2700, 2);
        coil.put(3600, 3);
        coil.put(4500, 4);
        coil.put(5400, 5);
        coil.put(7200, 6);
        coil.put(9001, 7);
        coil.put(10800, 8);
        coil.put(12600, 9);
        coil.put(14400, 10);
        coil.put(16200, 11);
        coil.put(18900, 12);
        coil.put(21600, 13);
        coil.put(36000, 14);
        coil.put(62000, 15);
        coil.put(96000, 16);
    }
}

File: src/main/java/com/gtladd/gtladditions/common/machine/muiltblock/controller/TaixuTurbidArray.kt
package com.gtladd.gtladditions.common.machine.muiltblock.controller

import com.gregtechceu.gtceu.api.GTCEuAPI
import com.gregtechceu.gtceu.api.GTValues
import com.gregtechceu.gtceu.api.block.ICoilType
import com.gregtechceu.gtceu.api.capability.recipe.FluidRecipeCapability
import com.gregtechceu.gtceu.api.capability.recipe.IO
import com.gregtechceu.gtceu.api.gui.GuiTextures
import com.gregtechceu.gtceu.api.machine.IMachineBlockEntity
import com.gregtechceu.gtceu.api.machine.MetaMachine
import com.gregtechceu.gtceu.api.machine.feature.IMachineModifyDrops
import com.gregtechceu.gtceu.api.machine.multiblock.WorkableMultiblockMachine
import com.gregtechceu.gtceu.api.machine.trait.NotifiableItemStackHandler
import com.gregtechceu.gtceu.api.recipe.GTRecipe
import com.gregtechceu.gtceu.api.recipe.RecipeHelper
import com.gregtechceu.gtceu.api.recipe.content.ContentModifier
import com.gregtechceu.gtceu.common.block.CoilBlock
import com.gregtechceu.gtceu.common.data.GTMaterials
import com.gregtechceu.gtceu.common.data.GTRecipeTypes
import com.gregtechceu.gtceu.data.recipe.builder.GTRecipeBuilder
import com.gtladd.gtladditions.GTLAdditions
import com.gtladd.gtladditions.common.items.GTLAddItems
import com.lowdragmc.lowdraglib.gui.widget.SlotWidget
import com.lowdragmc.lowdraglib.gui.widget.Widget
import com.lowdragmc.lowdraglib.gui.widget.WidgetGroup
import com.lowdragmc.lowdraglib.misc.ItemStackTransfer
import com.lowdragmc.lowdraglib.syncdata.annotation.Persisted
import com.lowdragmc.lowdraglib.syncdata.field.ManagedFieldHolder
import it.unimi.dsi.fastutil.ints.IntArrayList
import it.unimi.dsi.fastutil.objects.ObjectArrayList
import net.minecraft.MethodsReturnNonnullByDefault
import net.minecraft.network.chat.Component
import net.minecraft.world.item.Item
import net.minecraft.world.item.ItemStack
import net.minecraftforge.common.util.Lazy
import org.gtlcore.gtlcore.api.pattern.util.IValueContainer
import org.gtlcore.gtlcore.api.recipe.IParallelLogic
import org.gtlcore.gtlcore.common.data.GTLMaterials
import org.gtlcore.gtlcore.common.machine.multiblock.electric.TierCasingMachine
import org.gtlcore.gtlcore.utils.Registries
import java.text.DecimalFormat
import java.util.concurrent.ThreadLocalRandom
import javax.annotation.ParametersAreNonnullByDefault
import kotlin.math.*

@ParametersAreNonnullByDefault
@MethodsReturnNonnullByDefault
open class TaixuTurbidArray(holder: IMachineBlockEntity) : TierCasingMachine(holder, "SCTier"), IMachineModifyDrops {

    @field:Persisted
    val machineStorage: NotifiableItemStackHandler

    private var coilType: ICoilType = CoilBlock.CoilType.CUPRONICKEL
    private var height: Int = 0

    // structure
    private var frameA: Double = 0.0
    private var frameB: Double = 0.0
    private var uuAmplifierAmount: Long = 0
    private var uuMatterAmount: Long = 0

    // itemStack
    private var maxParallel: Long = 0
    private var successRateA: Double = 0.0
    private var successRateB: Double = 0.0

    init {
        this.machineStorage = createMachineStorage()
    }

    companion object {
        val MANAGED_FIELD_HOLDER: ManagedFieldHolder =
            ManagedFieldHolder(TaixuTurbidArray::class.java, WorkableMultiblockMachine.MANAGED_FIELD_HOLDER)

        // Constants - Frame calculations
        private const val FRAME_A_BASE = 8.0
        private const val FRAME_A_TIER_POWER = 2.0
        private const val FRAME_B_BASE = 3.8
        private const val FRAME_B_POWER_BASE = 1.3
        private const val FRAME_B_TEMP_DIVISOR = 36000.0
        private const val FRAME_B_TEMP_POWER = 0.7

        // Constants - Success rate calculations
        private const val MAX_SUCCESS_RATE = 100.0
        private const val SUCCESS_A_FACTOR = 0.1
        private const val SUCCESS_A_DIVISOR_A = 50.0
        private const val SUCCESS_A_DIVISOR_B = 100.0
        private const val SUCCESS_A_HEIGHT_DIVISOR = 9.0
        private const val SUCCESS_B_FACTOR = 0.02
        private const val SUCCESS_B_DIVISOR = 20.0
        private const val SUCCESS_B_HEIGHT_FACTOR = 7.0

        // Constants - Output calculations
        private const val OUTPUT_1_BASE = 4096
        private const val OUTPUT_1_FACTOR = 0.015
        private const val OUTPUT_1_HEIGHT_DIVISOR = 16.0
        private const val OUTPUT_2_BASE = 2250
        private const val OUTPUT_2_FACTOR = 0.06
        private const val OUTPUT_2_DIVISOR = 200.0

        // Constants - Parallel calculations
        private const val PARALLEL_BASE = 4096.0
        private const val PARALLEL_POWER_BASE = 1.621
        private const val PARALLEL_TEMP_DIVISOR = 6400.0
        private const val ASTRAL_BASE_PARALLEL = 43046721L // 3^16

        // Constants - Slot bonuses
        private const val ENDERIUM_BONUS = 0.01
        private const val DRACONIUM_BONUS = 0.05
        private const val SPACETIME_BONUS = 0.1
        private const val ETERNITY_BONUS = 0.2

        // Constants - Recipe
        private const val RECIPE_DURATION = 100
        private const val RECIPE_EU_MULTIPLIER = 524288L

        private val coil: Lazy<IntArrayList> = Lazy.of {
            IntArrayList(
                GTCEuAPI.HEATING_COILS.keys.stream()
                    .map { it.coilTemperature }
                    .sorted()
                    .toList()
            )
        }

        private val ENDERIUM_ITEM: ItemStack = Registries.getItemStack("gtceu:enderium_nanoswarm", 64)
        private val DRACONIUM_ITEM: ItemStack = Registries.getItemStack("gtceu:draconium_nanoswarm", 64)
        private val SPACETIME_ITEM: ItemStack = Registries.getItemStack("gtceu:spacetime_nanoswarm", 64)
        private val ETERNITY_ITEM: ItemStack = Registries.getItemStack("gtceu:eternity_nanoswarm", 64)
        private val ASTRAL_ARRAY_ITEM: ItemStack = GTLAddItems.ASTRAL_ARRAY.asStack()

        private val VALID_ITEMS: Set<Item> = setOf(
            ENDERIUM_ITEM.item,
            DRACONIUM_ITEM.item,
            SPACETIME_ITEM.item,
            ETERNITY_ITEM.item,
            ASTRAL_ARRAY_ITEM.item
        )

        // ==================== Recipe Modifier ====================
        fun recipeModifier(machine: MetaMachine, recipe: GTRecipe): GTRecipe? {
            if (machine !is TaixuTurbidArray) return null

            var modifiedRecipe = recipe.copy()
            val maxParallel = IParallelLogic.getMaxParallel(machine, recipe, machine.maxParallel)
            if (maxParallel <= 0) return null

            // Build fluid outputs based on success rates
            val builder = GTRecipeBuilder(GTLAdditions.id("uu"), GTRecipeTypes.DUMMY_RECIPES)
            modifiedRecipe.outputs.put(FluidRecipeCapability.CAP, ObjectArrayList())

            val random = ThreadLocalRandom.current()

            // UU Amplifier output (UXV+)
            if (machine.tier >= GTValues.UXV &&
                random.nextDouble(MAX_SUCCESS_RATE) <= machine.successRateA
            ) {
                builder.outputFluids(GTLMaterials.UuAmplifier.getFluid(machine.uuAmplifierAmount))
            }

            // UU Matter output (MAX+)
            if (machine.tier >= GTValues.MAX &&
                random.nextDouble(MAX_SUCCESS_RATE) <= machine.successRateB
            ) {
                builder.outputFluids(GTMaterials.UUMatter.getFluid(machine.uuMatterAmount))
            }

            // Add generated outputs to recipe
            val generatedOutputs = builder.buildRawRecipe().outputs[FluidRecipeCapability.CAP]
            if (generatedOutputs != null && generatedOutputs.isNotEmpty()) {
                modifiedRecipe.outputs[FluidRecipeCapability.CAP]!!.addAll(generatedOutputs)
            }

            // Apply parallel limit based on output merging
            val minParallel = IParallelLogic.getMinParallel(machine, modifiedRecipe, maxParallel)
            if (minParallel <= 0) return null

            // Apply parallel multiplier
            modifiedRecipe = modifiedRecipe.copy(ContentModifier.multiplier(minParallel.toDouble()), false)
            modifiedRecipe.duration = RECIPE_DURATION
            RecipeHelper.setInputEUt(modifiedRecipe, RECIPE_EU_MULTIPLIER * GTValues.V[machine.tier])

            return modifiedRecipe
        }
    }

    // ==================== Initialization ====================
    private fun createMachineStorage(): NotifiableItemStackHandler {
        val handler = NotifiableItemStackHandler(this, 1, IO.NONE, IO.BOTH) { _ ->
            object : ItemStackTransfer(1) {
                override fun onContentsChanged() {
                    recalculateItemCache()
                }
            }
        }
        handler.setFilter { itemStack -> filter(itemStack) }
        return handler
    }

    override fun onStructureFormed() {
        super.onStructureFormed()
        val context = this.multiblockState.matchContext

        val type = context.get<Any?>("CoilType")
        if (type is ICoilType) {
            this.coilType = type
        }

        val speedPipe = context.getOrCreate("SpeedPipeValue") { IValueContainer.noop() }.getValue()
        if (speedPipe is Int) {
            this.height = speedPipe - 2
        }

        recalculateAll()
    }

    private fun recalculateAll() {
        frameA = calculateFrameA()
        frameB = calculateFrameB()
        uuAmplifierAmount = calculateBaseOutputFluid1()
        uuMatterAmount = calculateBaseOutputFluid2()
        recalculateItemCache()
    }

    private fun recalculateItemCache() {
        val itemStack = machineStorage.storage.getStackInSlot(0)
        val item = itemStack.item
        val isAstralArray = ASTRAL_ARRAY_ITEM.`is`(item)

        val slotBonus = when {
            ENDERIUM_ITEM.`is`(item) -> ENDERIUM_BONUS * itemStack.count
            DRACONIUM_ITEM.`is`(item) -> DRACONIUM_BONUS * itemStack.count
            SPACETIME_ITEM.`is`(item) -> SPACETIME_BONUS * itemStack.count
            ETERNITY_ITEM.`is`(item) -> ETERNITY_BONUS * itemStack.count
            else -> 0.0
        }

        maxParallel = if (isAstralArray) {
            ASTRAL_BASE_PARALLEL * itemStack.count.toDouble().pow(5.0/3.0).toInt()
        } else {
            (PARALLEL_BASE * PARALLEL_POWER_BASE.pow(coilType.coilTemperature / PARALLEL_TEMP_DIVISOR)).toLong()
        }

        successRateA = if (isAstralArray) {
            MAX_SUCCESS_RATE
        } else {
            MAX_SUCCESS_RATE / (1 + exp(-SUCCESS_A_FACTOR *
                (frameA / SUCCESS_A_DIVISOR_A + frameB / SUCCESS_A_DIVISOR_B +
                height / SUCCESS_A_HEIGHT_DIVISOR))) + slotBonus
        }

        successRateB = if (isAstralArray) {
            MAX_SUCCESS_RATE
        } else {
            MAX_SUCCESS_RATE * (1 - exp(-SUCCESS_B_FACTOR *
                ((frameA + frameB) / SUCCESS_B_DIVISOR +
                cbrt(height.toDouble()) * tier / SUCCESS_B_HEIGHT_FACTOR))) + slotBonus
        }
    }

    // ==================== GUI ====================
    override fun createUIWidget(): Widget {
        val widget = super.createUIWidget()
        if (widget is WidgetGroup) {
            val size = widget.size
            widget.addWidget(
                SlotWidget(machineStorage.storage, 0, size.width - 30, size.height - 30, true, true)
                    .setBackground(GuiTextures.SLOT)
                    .setHoverTooltips(slotTooltips())
            )
        }
        return widget
    }

    private fun slotTooltips(): MutableList<Component> {
        return mutableListOf(
            Component.translatable("gtceu.machine.taixu.storage.tooltip.0"),
            Component.translatable("gtceu.machine.taixu.storage.tooltip.1"),
            Component.translatable("gtceu.machine.taixu.storage.tooltip.2"),
            Component.translatable("gtceu.machine.taixu.storage.tooltip.3"),
            Component.translatable("gtceu.machine.taixu.storage.tooltip.4"),
            Component.translatable("gtceu.machine.taixu.storage.tooltip.5"),
            Component.translatable("gtceu.machine.taixu.storage.tooltip.6")
        )
    }

    override fun addDisplayText(textList: MutableList<Component?>) {
        super.addDisplayText(textList)
        if (isFormed) {
            textList.add(Component.translatable("gtceu.machine.taixu.gui.tooltip.0", height))
            textList.add(Component.translatable("gtceu.machine.taixu.gui.tooltip.1", maxParallel))

            if (tier > GTValues.UIV) {
                val df = DecimalFormat(".00'%'")
                textList.add(Component.translatable("gtceu.machine.taixu.gui.tooltip.2", df.format(successRateA)))
                textList.add(Component.translatable("gtceu.machine.taixu.gui.tooltip.3", uuAmplifierAmount))

                if (tier > GTValues.OpV) {
                    textList.add(Component.translatable("gtceu.machine.taixu.gui.tooltip.4", df.format(
                        successRateB
                    )))
                    textList.add(Component.translatable("gtceu.machine.taixu.gui.tooltip.5", uuMatterAmount))
                }
            }
        }
    }

    // ==================== Logic ====================
    override fun beforeWorking(recipe: GTRecipe?): Boolean = true

    override fun onDrops(drops: MutableList<ItemStack?>) {
        clearInventory(machineStorage.storage)
    }

    private fun filter(itemStack: ItemStack): Boolean {
        return VALID_ITEMS.contains(itemStack.item)
    }

    // ==================== Calculation Methods ====================
    private fun calculateFrameA(): Double {
        return FRAME_A_BASE * (FRAME_A_TIER_POWER.pow(casingTier.toDouble()) - 1) *
               sqrt((GTValues.ALL_TIERS[tier] + 1).toDouble())
    }

    private fun calculateFrameB(): Double {
        val coilIndex = coil.get().indexOf(coilType.coilTemperature) + 1
        return FRAME_B_BASE * FRAME_B_POWER_BASE.pow(coilIndex.toDouble()) *
               (coilType.coilTemperature / FRAME_B_TEMP_DIVISOR).pow(FRAME_B_TEMP_POWER)
    }

    private fun calculateBaseOutputFluid1(): Long {
        return (OUTPUT_1_BASE * (1 - exp(-OUTPUT_1_FACTOR *
               (frameA * height / OUTPUT_1_HEIGHT_DIVISOR +
                frameB * ln((tier + 2).toDouble()))))).toLong()
    }

    private fun calculateBaseOutputFluid2(): Long {
        return (OUTPUT_2_BASE * tanh(sqrt(frameA * frameB) *
               (height + tier) * OUTPUT_2_FACTOR / OUTPUT_2_DIVISOR)).toLong()
    }

    override fun getFieldHolder(): ManagedFieldHolder = MANAGED_FIELD_HOLDER
}